test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py

```diff
@@ -0,0 +1,1473 @@
+import re
+
+import numpy as np
+import pytest
+from openqasm3.ast import (
+    ArrayLiteral,
+    ArrayType,
+    BitstringLiteral,
+    BitType,
+    BooleanLiteral,
+    FloatLiteral,
+    FloatType,
+    Identifier,
+    IndexedIdentifier,
+    IntegerLiteral,
+    IntType,
+    QuantumGate,
+    QuantumGateDefinition,
+    UintType,
+)
+
+from braket.default_simulator import StateVectorSimulation
+from braket.default_simulator.gate_operations import U
+from braket.default_simulator.openqasm import data_manipulation
+from braket.default_simulator.openqasm.circuit import Circuit
+from braket.default_simulator.openqasm.data_manipulation import convert_bool_array_to_string
+from braket.default_simulator.openqasm.interpreter import Interpreter
+from braket.default_simulator.openqasm.program_context import QubitTable
+
+
+def test_bit_declaration():
+    qasm = """
+    bit single_uninitialized;
+    bit single_initialized_int = 0;
+    bit single_initialized_bool = true;
+    bit[2] register_uninitialized;
+    bit[2] register_initialized = "01";
+    """
+    context = Interpreter().run(qasm)
+
+    assert context.get_type("single_uninitialized") == BitType(None)
+    assert context.get_type("single_initialized_int") == BitType(None)
+    assert context.get_type("single_initialized_bool") == BitType(None)
+    assert context.get_type("register_uninitialized") == BitType(IntegerLiteral(2))
+    assert context.get_type("register_initialized") == BitType(IntegerLiteral(2))
+
+    assert context.get_value("single_uninitialized") is None
+    assert context.get_value("single_initialized_int") == BooleanLiteral(False)
+    assert context.get_value("single_initialized_bool") == BooleanLiteral(True)
+    assert context.get_value("register_uninitialized") == ArrayLiteral([None, None])
+    assert context.get_value("register_initialized") == ArrayLiteral(
+        [BooleanLiteral(False), BooleanLiteral(True)]
+    )
+
+
+def test_int_declaration():
+    qasm = """
+    int[8] uninitialized;
+    int[8] pos = 10;
+    int[5] neg = -4;
+    int[3] pos_overflow = 5;
+    int[3] neg_overflow = -6;
+    """
+    pos_overflow = "Integer overflow for value 5 and size 3."
+    neg_overflow = "Integer overflow for value -6 and size 3."
+
+    with pytest.warns(UserWarning) as warn_info:
+        context = Interpreter().run(qasm)
+
+    assert context.get_type("uninitialized") == IntType(IntegerLiteral(8))
+    assert context.get_type("pos") == IntType(IntegerLiteral(8))
+    assert context.get_type("neg") == IntType(IntegerLiteral(5))
+    assert context.get_type("pos_overflow") == IntType(IntegerLiteral(3))
+    assert context.get_type("neg_overflow") == IntType(IntegerLiteral(3))
+
+    assert context.get_value("uninitialized") is None
+    assert context.get_value("pos") == IntegerLiteral(10)
+    assert context.get_value("neg") == IntegerLiteral(-4)
+    assert context.get_value("pos_overflow") == IntegerLiteral(1)
```

Comment:
OpenQASM uses a fixed point representation for its numbers, with a sign bit and magnitude bits, rather than the common two's complement representation [doc source](https://openqasm.com/language/types.html#classical-types:~:text=We%20use%20the%20notation%20s%3Am%3Af%20to%20denote%20the%20width%20and%20precision%20of%20fixed%20point%20numbers%2C%20where%20s%20is%20the%20number%20of%20sign%20bits%2C%20m%20is%20the%20number%20of%20integer%20bits%2C%20and%20f%20is%20the%20number%20of%20fractional%20bits.). Overflow is not specifically described in the spec, so it is generally accepted as implementation-specific.